crates/rpc/rpc-eth-api/src/helpers/call.rs

```diff
@@ -98,106 +98,57 @@ pub trait EthCall: EstimateCall + Call + LoadPendingBlock + LoadBlock + FullEthA
                     State::builder().with_database(StateProviderDatabase::new(state)).build();
                 let mut blocks: Vec<SimulatedBlock<RpcBlock<Self::NetworkTypes>>> =
                     Vec::with_capacity(block_state_calls.len());
+
+                let mut previous_block_number = None;
                 for block in block_state_calls {
-                    let mut evm_env = this
-                        .evm_config()
-                        .next_evm_env(&parent, &this.next_env_attributes(&parent)?)
-                        .map_err(RethError::other)
-                        .map_err(Self::Error::from_eth_err)?;
-
-                    // Always disable EIP-3607
-                    evm_env.cfg_env.disable_eip3607 = true;
-
-                    if !validation {
-                        evm_env.cfg_env.disable_base_fee = !validation;
-                        evm_env.block_env.basefee = 0;
-                    }
+                    if let Some(prev_block_number) = previous_block_number {
```

Comment:
I find this a bit hard to read.

I'd rather see a stack based approach where we can iterater over `block_state_calls` but also push new entries if there are gaps, or even compute this upfront

this should be a lot simpler than filling the gaps inside this loop, if we can push a new block_state_call or have `block_state_calls` sanitized from the start